executor/executor.go

```diff
@@ -98,36 +100,85 @@ func (e *TableScanExec) Fields() []*ast.ResultField {
 
 // Next implements Execution Next interface.
 func (e *TableScanExec) Next() (*Row, error) {
-	if e.iter == nil {
-		txn, err := e.ctx.GetTxn(false)
+	for {
+		if e.cursor >= len(e.ranges) {
+			return nil, nil
+		}
+		ran := e.ranges[e.cursor]
+		if e.seekHandle < ran.LowVal {
+			e.seekHandle = ran.LowVal
+		}
+		if e.seekHandle > ran.HighVal {
+			e.cursor++
+			continue
+		}
+		rowKey, err := e.seek()
+		if err != nil || rowKey == nil {
+			return nil, errors.Trace(err)
+		}
+		handle, err := tables.DecodeRecordKeyHandle(rowKey)
 		if err != nil {
 			return nil, errors.Trace(err)
 		}
-		e.iter, err = txn.Seek(e.t.FirstKey())
+		if handle > ran.HighVal {
+			// The handle is out of the current range, but may be in next range.
+			// We try to iterate to the range that contains the handle, so we
+			// don't need to seek again.
+			inRange := e.iterateToRangeContains(handle)
+			if !inRange {
+				continue
+			}
+		}
+		row, err := e.getRow(handle, rowKey)
 		if err != nil {
 			return nil, errors.Trace(err)
 		}
+		e.seekHandle = handle + 1
+		return row, nil
 	}
-	if !e.iter.Valid() || !e.iter.Key().HasPrefix(e.t.RecordPrefix()) {
-		return nil, nil
+}
+
+func (e *TableScanExec) seek() (kv.Key, error) {
+	seekKey := tables.EncodeRecordKey(e.t.TableID(), e.seekHandle, 0)
+	txn, err := e.ctx.GetTxn(false)
+	if err != nil {
+		return nil, errors.Trace(err)
 	}
-	// TODO: check if lock valid
-	// the record layout in storage (key -> value):
-	// r1 -> lock-version
-	// r1_col1 -> r1 col1 value
-	// r1_col2 -> r1 col2 value
-	// r2 -> lock-version
-	// r2_col1 -> r2 col1 value
-	// r2_col2 -> r2 col2 value
-	// ...
-	rowKey := e.iter.Key()
-	handle, err := tables.DecodeRecordKeyHandle(rowKey)
+	if e.iter != nil {
+		e.iter.Close()
+	}
+	e.iter, err = txn.Seek(seekKey)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
+	if !e.iter.Valid() || !e.iter.Key().HasPrefix(e.t.RecordPrefix()) {
+		// No more records in the table, skip to the end.
+		e.cursor = len(e.ranges)
+		return nil, nil
+	}
+	return e.iter.Key(), nil
+}
+
+func (e *TableScanExec) iterateToRangeContains(handle int64) (inRange bool) {
+	for {
+		e.cursor++
+		if e.cursor >= len(e.ranges) {
+			return false
+		}
+		ran := e.ranges[e.cursor]
+		if handle < ran.LowVal {
+			return false
+		}
+		if handle > ran.HighVal {
+			continue
+		}
+		return true
+	}
+}
 
-	// TODO: we could just fetch mentioned columns' values
+func (e *TableScanExec) getRow(handle int64, rowKey kv.Key) (*Row, error) {
 	row := &Row{}
+	var err error
 	row.Data, err = e.t.Row(e.ctx, handle)
 	if err != nil {
 		return nil, errors.Trace(err)
```

Comment:
Line 187 to line189 will be called multiply times? I think it is only necessary for the first row.